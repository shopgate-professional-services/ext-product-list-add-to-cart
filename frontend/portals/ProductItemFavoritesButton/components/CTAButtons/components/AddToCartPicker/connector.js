import { connect } from 'react-redux';
import {
  getProduct,
  hasProductVariety,
  isProductOrderable,
  getProductStock,
} from '@shopgate/pwa-common-commerce/product/selectors/product';
import addProductsToCart from '@shopgate/pwa-common-commerce/cart/actions/addProductsToCart';

const mapStateToProps = (state, props) => ({
  product: getProduct(state, props),
  isSimpleProduct: !hasProductVariety(state, props),
  isOrderable: isProductOrderable(state, props),
  stock: getProductStock(state, props),
});

const mapDispatchToProps = (dispatch, { productId }) => ({
  handleAddToCart: quantity => (
    dispatch(addProductsToCart([{ quantity, productId }]))
  ),
});

export default connect(mapStateToProps, mapDispatchToProps);
