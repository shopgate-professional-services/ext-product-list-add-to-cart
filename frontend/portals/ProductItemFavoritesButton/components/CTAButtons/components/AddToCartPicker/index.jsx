import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BasePicker from '@shopgate/pwa-common/components/Picker';
import Sheet from '@shopgate/pwa-ui-shared/Sheet';
import Conditioner from '@shopgate/pwa-core/classes/Conditioner';
import AddToCartPickerButton from './components/AddToCartPickerButton';
import List from './components/List';
import connect from './connector';
import createPickerItems from '../../../../../../helpers/createPickerItems';
import styles from './style';
import getConfig from '../../../../../../helpers/getConfig';

const { maxEntries } = getConfig();
const clickDelay = 150;

/**
 * AddToCartPicker Component
 */
class AddToCartPicker extends Component {
  static propTypes = {
    goToProductPage: PropTypes.func,
    handleAddToCart: PropTypes.func,
    isOrderable: PropTypes.bool,
    isSimpleProduct: PropTypes.bool,
    productName: PropTypes.string,
    showModal: PropTypes.func,
    stock: PropTypes.shape(),
  }

  static defaultProps = {
    goToProductPage: () => { },
    handleAddToCart: () => { },
    isOrderable: true,
    isSimpleProduct: false,
    productName: null,
    showModal: () => Promise().resolve(true),
    stock: null,
  };

  static contextTypes = {
    i18n: PropTypes.func,
  };

  /**
   * Component Constructor
   * @param {Object} props Component props
   */
  constructor(props) {
    super(props);
    this.productConditioner = new Conditioner();
    this.productConditioner.addConditioner('validateProduct', this.validateProduct);
  }

  state = {
    addedQuantity: 0,
  };

  /**
   * Returns the props for the picker button.
   * @return {Object}
   */
  get buttonProps() {
    let props = {
      isDisabled: !this.props.isOrderable,
      conditioner: this.productConditioner,
      addedQuantity: this.state.addedQuantity,
      handleAddToCart: /* istanbul ignore next */ () => { },
      onClick: () => { },
      isLoading: false,
      hasLoading: true,
      noShadow: false,
      className: null,
      type: 'default',
    };

    if (props.noShadow) {
      // Inject additional classes when the button is rendered without a shadow.
      props = {
        ...props,
        className: props.isDisabled ? styles.addToCartShadow : styles.noShadow,
      };
    }
    return props;
  }

  /**
   * Provides product validation for product conditioner
   * @return {boolean}
   */
  validateProduct = () => {
    const {
      isOrderable, isSimpleProduct, showModal, goToProductPage,
    } = this.props;
    if (!isOrderable) {
      return false;
    }
    if (!isSimpleProduct) {
      showModal({
        message: 'product_list_add_to_cart.modal.message',
        confirm: 'product_list_add_to_cart.modal.confirm',
        dismiss: 'product_list_add_to_cart.modal.cancel',
      })
        .then((result) => {
          if (result) {
            goToProductPage();
          }
        });
      return false;
    }
    return true;
  }

  /**
   * @returns {JSX}
   */
  listComponent = ({ items, onSelect, onClose }) => (
    <List>
      {items.map(item => (
        <List.Item
          key={item.value}
          title={item.label}
          onClick={() => {
            setTimeout(() => {
              onSelect(item.value);
              onClose();
            }, clickDelay);
          }}
        />
      ))}
    </List>
  );

  /**
   * @param {Object} modalProps Props for modal
   * @returns {JSX}
   */
  modalComponent = (modalProps) => {
    const { __ } = this.context.i18n();
    const translatedTitle = __('product_list_add_to_cart.sheet_title');
    return (<Sheet {...modalProps} title={`${translatedTitle} ${this.props.productName}`} />);
  }

  /**
   * Function to handle cart
   * @param {number} quantity quantity selected.
   */
  handelAddToCart = (quantity) => {
    this.props.handleAddToCart(quantity);

    this.setState(prevState => ({
      addedQuantity: prevState.addedQuantity + quantity,
    }));
  };

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const pickerItems = createPickerItems(this.props.stock, maxEntries);
    return (
      <BasePicker
        modalComponent={this.modalComponent}
        buttonProps={this.buttonProps}
        buttonComponent={AddToCartPickerButton}
        items={pickerItems}
        listComponent={this.listComponent}
        onSelect={this.handelAddToCart}
      />
    );
  }
}

export default connect(AddToCartPicker);
