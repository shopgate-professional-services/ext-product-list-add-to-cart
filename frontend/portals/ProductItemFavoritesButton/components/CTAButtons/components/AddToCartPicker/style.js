import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { shadows } = themeConfig;

const addToCartShadow = css({
  display: 'absolute',
  boxShadow: shadows.buttons.elevated,
}).toString();

const noShadow = css({
  display: 'absolute',
  boxShadow: shadows.buttons.disabled,
});

export default {
  addToCartShadow,
  noShadow,
};
