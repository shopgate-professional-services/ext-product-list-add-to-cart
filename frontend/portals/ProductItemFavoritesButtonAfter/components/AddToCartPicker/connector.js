import { connect } from 'react-redux';
import {
  getProduct,
  hasProductVariety,
  isProductOrderable,
  getProductStock,
} from '@shopgate/pwa-common-commerce/product/selectors/product';

const mapStateToProps = (state, props) => ({
  product: getProduct(state, props),
  isSimpleProduct: !hasProductVariety(state, props),
  isOrderable: isProductOrderable(state, props),
  stock: getProductStock(state, props),
});

export default connect(mapStateToProps);
