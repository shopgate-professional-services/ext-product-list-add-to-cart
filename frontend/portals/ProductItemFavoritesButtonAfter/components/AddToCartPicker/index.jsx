import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BasePicker from '@shopgate/pwa-common/components/Picker';
import Sheet from '@shopgate/pwa-ui-shared/Sheet';
import List from './components/List';
import AddToCartButton from './components/AddToCartButton';
import connect from './connector';
import getConfig from '../../../../helpers/getConfig';
import createPickerItems from '../../../../helpers/createPickerItems';

/**
 * The AddToCartPicker component.
 */
class AddToCartPicker extends Component {

  state = {
    addedQuantity: 0,
  };

  listComponent = ({ items, onSelect }) => (
    <List>
      {items.map(item => (
        <List.Item
          key={item.value}
          title={item.label}
          onClick={() => {
            setTimeout(() => {
              onSelect(item.value);
            }, 50);
          }}
        />
      ))}
    </List>
  );


  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const pickerItems = createPickerItems(this.props.stock);
    return (
      <BasePicker
        onClick={()=>{}}
        items={pickerItems}
        modalComponent={() => (<Sheet title="Aaron is amazing" />)}
        buttonProps={{}}
        buttonComponent={AddToCartButton}
        listComponent={this.listComponent}
        onSelect={()=>{}}
      />
    );
  }
}

export default connect(AddToCartPicker);
