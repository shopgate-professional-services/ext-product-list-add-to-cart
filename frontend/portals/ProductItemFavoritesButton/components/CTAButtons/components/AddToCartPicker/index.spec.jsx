import React from 'react';
import { shallow } from 'enzyme';
import { createMockStore } from '@shopgate/pwa-common/store';
import { Provider } from 'react-redux';
import AddToCartPicker from './index';

const store = createMockStore();

describe('AddToCartPicker', () => {
  it('should render', () => {
    const component = shallow((
      <Provider store={store}>
        <AddToCartPicker />
      </Provider>
    ));
    expect(component).toMatchSnapshot();
  });
});
