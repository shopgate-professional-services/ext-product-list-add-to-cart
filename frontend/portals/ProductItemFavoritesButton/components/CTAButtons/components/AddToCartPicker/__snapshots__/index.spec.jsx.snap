// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToCartPicker should render 1`] = `<Connect(AddToCartPicker) />`;
